### PR TITLE
RELEASE_ASSERT on insanely large headers (rather than allow overflow attacks)

### LINKED ISSUE #1421: harden header map overflow code
(This is a potential security issue but exploiting is very difficult and may be impossible so just opening issue).

```
Hi,

while looking at Envoy's code, I noticed that in header_map_impl.cc's HeaderString::append and setCopy the calculation of the capacity of the buffer is:

      uint32_t new_capacity = (string_length_ + size) * 2;
      buffer_.dynamic_ = static_cast<char*>(malloc(new_capacity));
      memcpy(buffer_.dynamic_, inline_buffer_, string_length_);

size is the parameter to the function and of type uint32_t. It is the size of the header value that gets copied into HeaderString.

If size is more than 2GB, the calculation of new_capacity can wrap and the allocation will be too small. Subsequently, an out-of-bounds heap write will occur during the memcpy.

I also recommend to check if malloc returns !NULL to prevent null pointer dereference under memory pressure.

Transferring 2GB could be a tough attack scenario but with HTTP2 compression it might be easier. I didn't invest time to do a proof-of-concept. Let me know if you need further information.

If you want to attribute the vulnerability please attribute to Felix Groebert, Google Security Team.

Thank you
Felix
```

In HTTP/1 hitting an overflow is definitely impossible due to max header block size tracking in http_parser.

For HTTP/2, we should add additional checking here: https://github.com/lyft/envoy/blob/master/source/common/http/http2/codec_impl.cc#L680 to make sure an individual header does not go above max size.

In the header map code itself we should probably add comments and debug asserts to check against overflow. If we want to be paranoid we can add RELEASE_ASSERT there but not a big fan of that if we think it is impossible.

Assigning over to @alyssawilk for investigation/fix.

### PR DESCRIPTION
**Pull request by alyssawilk**
Verified via code analysis that both HTTP and H2 code paths guard against incremental large headers well before we could allocate crazy large headers.    Added regression tests which would likely catch that if it changed and closed the overflow attack with a RELEASE_ASSERT.

Fixes #1421

Between the potentially-avoidable-but-not-really-worth-it RELEASE_ASSERT and writing code using exceptions this finalizes my transition from GFE team to Envoy team.  There's no going back now!

### CODE REVIEW

File: source/common/http/header_map_impl.cc [hunk 1]
```diff
@@ -86,7 +86,10 @@ void HeaderString::append(const char* data, uint32_t size) {
   case Type::Dynamic: {
     // We can get here either because we didn't fit in inline or we are already dynamic.
     if (type_ == Type::Inline) {
-      uint32_t new_capacity = (string_length_ + size) * 2;
+      uint64_t new_capacity = (static_cast<uint64_t>(string_length_) + size) * 2;
```
Comment (htuch) [inline source/common/http/header_map_impl.cc]: Nit: `const`

File: source/common/http/header_map_impl.cc [hunk 2]
```diff
@@ -86,7 +86,10 @@ void HeaderString::append(const char* data, uint32_t size) {
...
+      // If the resizing will cause buffer overflow due to hitting uint32_t::max, an OOM is likely
+      // imminent.  Fast-fail rather than allow a buffer overflow attack (issue #1421)
+      RELEASE_ASSERT(new_capacity <= std::numeric_limits<uint32_t>::max());
```
Comment (htuch) [inline source/common/http/header_map_impl.cc]: I was going to suggest a death test, but this probably will break CI ;)

File: test/common/http/http1/codec_impl_test.cc [hunk 3]
```diff
@@ -728,6 +728,35 @@ TEST_F(Http1ClientConnectionImplTest, WatermarkTest) {
...
+  std::string long_string = "foo: " + std::string(1024, 'q') + "\r\n";
+  for (int i = 0; i < 79; ++i) {
+    buffer = Buffer::OwnedImpl(long_string);
+    codec_->dispatch(buffer);
+  }
+  try {
+    buffer = Buffer::OwnedImpl(long_string);
+    codec_->dispatch(buffer);
+  } catch (const Envoy::EnvoyException& ex) {
+    exception_reason = ex.what();
```
Comment (htuch) [inline test/common/http/http1/codec_impl_test.cc]: What happens on the data path when we take an exception? Most exceptions in Envoy that I'm familiar with today happen on the config update path.

Comment (mattklein123) [reply test/common/http/http1/codec_impl_test.cc]: Protocol errors are raised as exceptions.

Comment (mattklein123) [reply test/common/http/http1/codec_impl_test.cc]: P.S. You can use EXPECT_THROW_WITH_MESSAGE here.

**Comment by mattklein123** [comment]
@alyssawilk LGTM other than comments. Can you merge master to pickup Circle tests?

File: test/common/http/http1/codec_impl_test.cc (hunk 3 above)
Comment (alyssawilk) [reply test/common/http/http1/codec_impl_test.cc]: Ah, I didn't know about that one, thanks!
(I'm totally claiming the macro means I haven't written exception code yet.  heh heh heh)

**Comment by mattklein123** [comment]
Sorry will need another master merge here to fix release test.

**Pull request merged**